File: iptfs/mbuf.py
# -*- coding: utf-8 eval: (yapf-mode 1) -*-
#
# January 22 2019, Christian Hopps <chopps@gmail.com>
#
# Copyright (c) 2019, LabN Consulting, L.L.C.
# All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
from __future__ import absolute_import, division, unicode_literals, print_function, nested_scopes

import logging
import threading

logger = logging.getLogger(__file__)


class MBuf:
    def __init__(self, size, hdrspace, refcnt=False):
        self.space = memoryview(bytearray(size))
        self.reset(hdrspace)
        self.end = self.start = self.space[hdrspace:]
        self.seq = self.flags = 0
        self.reflock = None
        if refcnt:
            self.reflock = threading.Lock()
            self.refcnt = 0

    def reset(self, hdrspace):
        self.end = self.start = self.space[hdrspace:]
        self.flags = self.seq = 0

    def addref(self):
        with self.reflock:
            self.refcnt += 1
            return self.refcnt

    def deref(self, freeq):
        with self.reflock:
            self.refcnt -= 1
            if self.refcnt != 0:
                return False
        freeq.push(self, True)
        return True

    def after(self):
        return self.end.nbytes

    def len(self):
        return self.start.nbytes - self.end.nbytes


class MQueue:
    def __init__(self, name, count, maxbuf, hdrspace, refcnt, debug):  # pylint: disable=R0913
        """MQueue is a queue for MBUfs.

        If maxbuf is non-0 then the queue will allocate and push count empty
        mbufs on creation.
        """

        self.name = name
        self.mcount = count
        self.maxbuf = maxbuf
        self.manage = self.maxbuf != 0
        self.hdrspace = hdrspace
        self.debug = debug

        self.lock = threading.Lock()
        self.push_cv = threading.Condition(self.lock)
        self.pop_cv = threading.Condition(self.lock)
        self.mbufs = []
        if self.manage:
            for _ in range(0, count):
                self.mbufs.append(MBuf(maxbuf, hdrspace, refcnt))

    def empty(self):
        return len(self.mbufs) == 0

    def full(self):
        return len(self.mbufs) >= self.mcount

    def pop(self):
        with self.pop_cv:
            while self.empty():
                if self.debug:
                    logger.debug("pop: mqueue %s is empty", self.name)
                self.pop_cv.wait()

            # If we were full then notify there will be push space.
            # if self.full():
            #     self.push_cv.notify()
            self.push_cv.notify()
            return self.mbufs.pop()

    def trypop(self):
        with self.pop_cv:
            if self.empty():
                return None

            # If we were full then notify there will be push space.
            if self.full():
                self.push_cv.notify()
            return self.mbufs.pop()

    def push(self, m, reset=False):
        """push an mbuf on the queue.

        If reset is true then the mbuf is reset to an initial state.
        """
        if reset:
            m.seq = 0
            m.reset(self.hdrspace)

        with self.push_cv:
            while self.full():
                if self.debug:
                    logger.debug("push: queue %s is full", self.name)
                self.push_cv.wait()

            # If we were empty then notify there will be something to pop.
            # if self.empty():
            #     self.pop_cv.notify()
            self.pop_cv.notify()
            self.mbufs.append(m)


class MIOVBuf:
    def __init__(self):
        self.mbufs = []
        self.iov = []
        self.mlen = 0

    def addmbuf(self, m, start):
        m.addref()
        self.mbufs.append(m)
        self.iov.append(start)
        self.mlen += len(start)

    def reset(self, freeq):
        for m in self.mbufs:
            m.deref(freeq)
        self.mbufs = []
        self.iov = []
        self.mlen = 0

    def len(self):
        return self.mlen


class MIOVQ:
    def __init__(self, name, size, freeq=None, debug=False):
        """MIOVQ is a queue for MIOVBufs.

        :Parameters:
            - `name` (`src`) - descriptive name for the queue.
            - `size` (`int`) - max depth of the queue.
            - `freeq` (`MQueue`) - queue to free refered mbufs into.
            - `debug` (`bool`) - enable debug logging.

        If freeq is not None then the queue will allocate and push count empty
        miovbufs on creation.
        """

        self.name = name
        self.mcount = size
        self.freeq = freeq
        self.debug = debug
        self.manage = freeq is not None

        self.lock = threading.Lock()
        self.push_cv = threading.Condition(self.lock)
        self.pop_cv = threading.Condition(self.lock)
        self.queue = []
        if self.manage:
            for _ in range(0, size):
                self.queue.append(MIOVBuf())

    def empty(self):
        return len(self.queue) == 0

    def full(self):
        return len(self.queue) >= self.mcount

    def pop(self):
        with self.pop_cv:
            while self.empty():
                if self.debug:
                    logger.debug("pop: mqueue %s is empty", self.name)
                self.pop_cv.wait()
            self.push_cv.notify()
            return self.queue.pop()

    def trypop(self):
        with self.pop_cv:
            if self.empty():
                return None
            self.push_cv.notify()
            return self.queue.pop()

    def push(self, m):
        """push an MIOVBuf on the queue.

        If this is a free-ing queue then reset the MIOVBuf.
        """
        if self.freeq is not None:
            m.reset(self.freeq)

        with self.push_cv:
            while self.full():
                if self.debug:
                    logger.debug("push: queue %s is full", self.name)
                self.push_cv.wait()

            self.pop_cv.notify()
            self.queue.append(m)


__author__ = 'Christian Hopps'
__date__ = 'January 22 2019'
__version__ = '1.0'
__docformat__ = "restructuredtext en"

File: src/mbuf.c
/*
 * -*- coding: utf-8 -*-*
 *
 * February 1 2019, Christian E. Hopps <chopps@labn.net>
 *
 * Copyright (c) 2019, LabN Consulting, L.L.C.
 * All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include "iptfs.h"
#include <err.h>
#include <pthread.h>
#include <string.h>
#include <sys/uio.h>

/*
 * mbuFs
 */

struct mbuf *
mbuf_new(size_t maxbuf, size_t hdrspace)
{
	struct mbuf *m = xzmalloc(maxbuf + sizeof(*m));
	m->space = (void *)&m[1];
	m->espace = m->space + maxbuf;
	m->end = m->start = &m->space[hdrspace];
	return m;
}

struct mqueue {
	const char *name;
	struct mbuf **queue;
	int size;
	int depth;
	ssize_t hdrspace;
	pthread_mutex_t lock;
	pthread_cond_t pushcv;
	pthread_cond_t popcv;
	struct ackinfo ackinfo;
};

struct miovq {
	const char *name;
	struct mqueue *freeq;
	struct miov **queue;
	int size;
	int depth;
	pthread_mutex_t lock;
	pthread_cond_t pushcv;
	pthread_cond_t popcv;
};

/*
 * mqueues - mbuf queues
 */

#define MQ_FULL(mq) ((mq)->depth == (mq)->size)
#define MQ_EMPTY(mq) ((mq)->depth == 0)

struct mqueue *
mqueue_new(const char *name, int size)
{
	struct mqueue *mq;
	if ((mq = malloc(sizeof(struct mqueue))) == NULL)
		err(1, "mqueue_new: malloc mqueue");
	if ((mq->queue = malloc(sizeof(struct mbuf *) * size)) == NULL)
		err(1, "mqueue_new: malloc queue");

	memset(&mq->ackinfo, 0, sizeof(mq->ackinfo));
	mq->name = name;
	mq->size = size;
	mq->depth = 0;

	pthread_mutex_init(&mq->lock, NULL);
	pthread_cond_init(&mq->pushcv, NULL);
	pthread_cond_init(&mq->popcv, NULL);

	return mq;
}

struct mqueue *
mqueue_new_freeq(const char *name, int size, ssize_t maxbuf, ssize_t hdrspace)
{
	struct mqueue *mq = mqueue_new(name, size);
	struct mbuf *mbufs;
	uint8_t *space;
	int i, bi;

	mq->hdrspace = hdrspace;

	if ((space = malloc(size * maxbuf)) == NULL)
		err(1, "mqueue_new_freeq: malloc space");
	if ((mbufs = malloc(size * sizeof(struct mbuf))) == NULL)
		err(1, "mqueue_new_freeq: malloc mbufs");

	for (bi = 0, i = 0; i < size; i++, bi += maxbuf) {
		struct mbuf *m = &mbufs[i];
		m->space = &space[bi];
		m->espace = m->space + maxbuf;
		m->start = &m->space[hdrspace];
		m->end = m->start;
		mq->queue[i] = m;
		mq->depth++;
	}

	return mq;
}

struct mbuf *
mqueue_pop(struct mqueue *mq)
{
	struct mbuf *m;

	pthread_mutex_lock(&mq->lock);
	while (MQ_EMPTY(mq)) {
		DBG("mqueue_pop: %s empty\n", mq->name);
		pthread_cond_wait(&mq->popcv, &mq->lock);
	}
	pthread_cond_signal(&mq->pushcv);
	m = mq->queue[--mq->depth];
	pthread_mutex_unlock(&mq->lock);

	return m;
}

void
mqueue_wait(struct mqueue *mq)
{
	struct mbuf *m;

	pthread_mutex_lock(&mq->lock);
	while (MQ_EMPTY(mq)) {
		pthread_cond_wait(&mq->popcv, &mq->lock);
	}
	pthread_mutex_unlock(&mq->lock);
	return;
}
struct mbuf *
mqueue_trypop(struct mqueue *mq, int maxsize)
{
	struct mbuf *m;

	pthread_mutex_lock(&mq->lock);
	if (MQ_EMPTY(mq)) {
		m = NULL;
		goto out;
	}
	pthread_cond_signal(&mq->pushcv);
	m = mq->queue[--mq->depth];
	if (maxsize && MBUF_LEN(m) > maxsize) {
		m = NULL;
		goto out;
	}
	--mq->depth;
out:
	pthread_mutex_unlock(&mq->lock);
	return m;
}

int
mqueue_push(struct mqueue *mq, struct mbuf *m, bool reset)
{
	int depth;
	if (reset)
		mbuf_reset(m, mq->hdrspace);

	pthread_mutex_lock(&mq->lock);
	while (MQ_FULL(mq)) {
		DBG("mqueue_push: %s full\n", mq->name);
		pthread_cond_wait(&mq->pushcv, &mq->lock);
	}
	pthread_cond_signal(&mq->popcv);
	mq->queue[mq->depth++] = m;
	depth = mq->depth;
	pthread_mutex_unlock(&mq->lock);
	return depth;
}

void
mqueue_get_ackinfo(struct mqueue *mq, uint32_t *drops, uint32_t *start,
		   uint32_t *end)
{
	pthread_mutex_lock(&mq->lock);
	*drops = mq->ackinfo.ndrop;
	*start = mq->ackinfo.start;
	*end = mq->ackinfo.last;
	memset(&mq->ackinfo, 0, sizeof(mq->ackinfo));
	pthread_mutex_unlock(&mq->lock);
}

struct ackinfo *
mqueue_get_ackinfop(struct mqueue *mq)
{
	return &mq->ackinfo;
}

struct miovq *
miovq_new(const char *name, int size)
{
	struct miovq *mq;
	if ((mq = xzmalloc(sizeof(struct miovq))) == NULL)
		err(1, "miovq_new: malloc miovq");
	if ((mq->queue = malloc(sizeof(struct miov *) * size)) == NULL)
		err(1, "miovq_new: malloc queue");
	mq->name = name;
	mq->size = size;

	pthread_mutex_init(&mq->lock, NULL);
	pthread_cond_init(&mq->pushcv, NULL);
	pthread_cond_init(&mq->popcv, NULL);

	return mq;
}

struct miovq *
miovq_new_freeq(const char *name, int size, int maxiov, struct mqueue *freeq)
{
	struct miovq *mq = miovq_new(name, size);
	uint8_t *space;
	int i, sz;

	mq->freeq = freeq;
	sz = sizeof(struct miov) +
	     (sizeof(struct iovec) + sizeof(struct mbuf *)) * maxiov;
	if ((space = malloc(size * sz)) == NULL)
		err(1, "miovq_new_freeq: malloc miov");

	for (i = 0; i < size; i++) {
		struct miov *m = (struct miov *)(space + i * sz);
		m->iov = (struct iovec *)(m + 1);
		m->mbufs = (struct mbuf **)(m->iov + maxiov);
		m->maxiov = maxiov;
		mq->queue[i] = m;
		mq->depth++;
	}

	return mq;
}

struct miov *
miovq_pop(struct miovq *mq)
{
	/* XXX an exact copy of mqueue_pop except for the types.. sigh, C */
	struct miov *m;

	pthread_mutex_lock(&mq->lock);
	while (MQ_EMPTY(mq)) {
		DBG("miovq_pop: %s empty\n", mq->name);
		pthread_cond_wait(&mq->popcv, &mq->lock);
	}
	pthread_cond_signal(&mq->pushcv);
	m = mq->queue[--mq->depth];
	pthread_mutex_unlock(&mq->lock);

	return m;
}

void
miov_reset(struct miov *m, struct miovq *mq)
{
	int niov = m->niov;
	for (int i = 0; i < niov; i++)
		mbuf_deref(mq->freeq, m->mbufs[i]);
	m->niov = 0;
	m->len = 0;
	m->left = 0;
}

int
miovq_push(struct miovq *mq, struct miov *m)
{
	int depth;
	if (mq->freeq)
		miov_reset(m, mq);

	pthread_mutex_lock(&mq->lock);
	while (MQ_FULL(mq)) {
		DBG("miovq_push: %s full\n", mq->name);
		pthread_cond_wait(&mq->pushcv, &mq->lock);
	}

	pthread_cond_signal(&mq->popcv);
	mq->queue[mq->depth++] = m;
	depth = mq->depth;
	pthread_mutex_unlock(&mq->lock);

	return depth;
}

/* Local Variables: */
/* c-file-style: "bsd" */
/* c-c++-enable-clang-format-on-save: t */
/* End: */
